File: projects/gnomad-api/src/schema/datasets/shared/flags.js
export const getBaseFlags = variantData => {
  const flags = []

  // gnomAD v3 variants have an lcr flag at the top level.
  // gnomAD v2.1 variants have lcr and segdup flags nested under a `flags` object.
  if (variantData.lcr || (variantData.flags && variantData.flags.lcr)) {
    flags.push('lcr')
  }

  if (variantData.flags && variantData.flags.segdup) {
    flags.push('segdup')
  }

  return flags
}

const LOF_CONSEQUENCE_TERMS = new Set([
  'transcript_ablation',
  'splice_acceptor_variant',
  'splice_donor_variant',
  'stop_gained',
  'frameshift_variant',
])

// "NC" was removed from the data pipeline some time ago, but some ExAC variants still have it.
const isLofOnNonCodingTranscript = transcriptConsequence =>
  (LOF_CONSEQUENCE_TERMS.has(transcriptConsequence.major_consequence) &&
    !transcriptConsequence.lof) ||
  transcriptConsequence.lof === 'NC'

const getFlagsForGeneContext = (variantData, geneId) => {
  const flags = getBaseFlags(variantData)

  const allConsequences =
    variantData.sorted_transcript_consequences || variantData.sortedTranscriptConsequences || []
  const consequencesInGene = allConsequences.filter(csq => csq.gene_id === geneId)
  const lofConsequencesInGene = consequencesInGene.filter(csq => csq.lof)
  const mostSevereConsequenceInGene = consequencesInGene[0]

  if (lofConsequencesInGene.length > 0) {
    // In some cases, a coding transcript with a non-pLoF VEP annotation may be ranked higher
    // than a non-coding transcript with a pLoF VEP annotation. LOFTEE does not annotate
    // non-coding transcripts. Check for mostSevereConsequenceInGene.lof to avoid showing an
    // LC pLoF flag next to a non-pLoF VEP consequence.
    if (!lofConsequencesInGene.some(csq => csq.lof === 'HC') && mostSevereConsequenceInGene.lof) {
      flags.push('lc_lof')
    }

    if (lofConsequencesInGene.every(csq => csq.lof_flags)) {
      flags.push('lof_flag')
    }
  }

  if (mostSevereConsequenceInGene) {
    // See gnomadjs#364
    if (isLofOnNonCodingTranscript(mostSevereConsequenceInGene)) {
      flags.push('nc_transcript')
    }
    if (mostSevereConsequenceInGene.lof === 'OS') {
      flags.push('os_lof')
    }
  }

  return flags
}

const getFlagsForRegionContext = variantData => {
  const flags = getBaseFlags(variantData)

  const allConsequences =
    variantData.sorted_transcript_consequences || variantData.sortedTranscriptConsequences || []
  const lofConsequences = allConsequences.filter(csq => csq.lof)
  const mostSevereConsequence = allConsequences[0]

  if (lofConsequences.length > 0) {
    // In some cases, a coding transcript with a non-pLoF VEP annotation may be ranked higher
    // than a non-coding transcript with a pLoF VEP annotation. LOFTEE does not annotate
    // non-coding transcripts. Check for mostSevereConsequence.lof to avoid showing an
    // LC pLoF flag next to a non-pLoF VEP consequence.
    if (!lofConsequences.some(csq => csq.lof === 'HC') && mostSevereConsequence.lof) {
      flags.push('lc_lof')
    }

    if (lofConsequences.every(csq => csq.lof_flags)) {
      flags.push('lof_flag')
    }
  }

  if (mostSevereConsequence) {
    if (isLofOnNonCodingTranscript(mostSevereConsequence)) {
      flags.push('nc_transcript')
    }
    if (mostSevereConsequence.lof === 'OS') {
      flags.push('os_lof')
    }
  }

  return flags
}

const getFlagsForTranscriptContext = (variantData, transcriptId) => {
  const flags = getBaseFlags(variantData)

  const allConsequences =
    variantData.sorted_transcript_consequences || variantData.sortedTranscriptConsequences || []
  const consequenceInTranscript = allConsequences.find(csq => csq.transcript_id === transcriptId)

  if (consequenceInTranscript) {
    if (consequenceInTranscript.lof === 'LC') {
      flags.push('lc_lof')
    }

    if (consequenceInTranscript.lof === 'OS') {
      flags.push('os_lof')
    }

    if (consequenceInTranscript.lof && consequenceInTranscript.lof_flags) {
      flags.push('lof_flag')
    }

    if (isLofOnNonCodingTranscript(consequenceInTranscript)) {
      flags.push('nc_transcript')
    }
  }

  return flags
}

export const getFlagsForContext = context => {
  switch (context.type) {
    case 'gene':
      return variantData => getFlagsForGeneContext(variantData, context.geneId)
    case 'region':
      return getFlagsForRegionContext
    case 'transcript':
      return variantData => getFlagsForTranscriptContext(variantData, context.transcriptId)
    default:
      throw Error(`Invalid context for getFlags: ${context.type}`)
  }
}

File: data/data_utils/computed_fields/flags.py
import hail as hl


def get_expr_for_consequence_lc_lof_flag(transcript_consequence):
    """Flag a transcript consequence if it has an LOFTEE annotation other than HC"""
    return hl.or_else((transcript_consequence.lof != "") & (transcript_consequence.lof != "HC"), False)


def get_expr_for_variant_lc_lof_flag(sorted_transcript_consequences):
    """Flag a variant if it has some transcript consequences with LOFTEE annotations and none are marked HC"""
    return hl.bind(
        lambda lof_annotations: (lof_annotations.size() > 0) & lof_annotations.all(lambda csq: csq.lof != "HC"),
        sorted_transcript_consequences.filter(lambda csq: csq.lof != ""),
    )


def get_expr_for_genes_with_lc_lof_flag(sorted_transcript_consequences):
    """
    From a variant's sorted transcript consequences, get the set of gene IDs where the variant has at least one
    LoF consequence in that gene and all the variant's LoF consequences in that gene are not marked HC.
    """
    return hl.bind(
        lambda lof_consequences: hl.set(lof_consequences.map(lambda csq: csq.gene_id)).filter(
            lambda gene_id: hl.bind(
                lambda lof_consequences_in_gene: (lof_consequences_in_gene.size() > 0)
                & (lof_consequences_in_gene.all(lambda csq: csq.lof != "HC")),
                lof_consequences.filter(lambda csq: csq.gene_id == gene_id),
            )
        ),
        sorted_transcript_consequences.filter(lambda csq: csq.lof != ""),
    )


def get_expr_for_consequence_loftee_flag_flag(transcript_consequence):
    """Flag a transcript consequence if it has a LOFTEE annotation with flags"""
    return hl.or_else((transcript_consequence.lof != "") & (transcript_consequence.lof_flags != ""), False)


def get_expr_for_variant_loftee_flag_flag(sorted_transcript_consequences):
    """Flag a variant if it has some transcript consequences with LOFTEE annotations and all have flags"""
    return hl.bind(
        lambda lof_annotations: (lof_annotations.size() > 0) & lof_annotations.all(lambda csq: csq.lof_flags != ""),
        sorted_transcript_consequences.filter(lambda csq: csq.lof != ""),
    )


def get_expr_for_genes_with_loftee_flag_flag(sorted_transcript_consequences):
    """
    From a variant's sorted transcript consequences, get the set of gene IDs where the variant has at least one
    LoF consequence in that gene and all the variant's LoF consequences in that gene are flagged by LOFTEE.
    """
    return hl.bind(
        lambda lof_consequences: hl.set(lof_consequences.map(lambda csq: csq.gene_id)).filter(
            lambda gene_id: hl.bind(
                lambda lof_consequences_in_gene: (lof_consequences_in_gene.size() > 0)
                & (lof_consequences_in_gene.all(lambda csq: csq.lof_flags != "")),
                lof_consequences.filter(lambda csq: csq.gene_id == gene_id),
            )
        ),
        sorted_transcript_consequences.filter(lambda csq: csq.lof != ""),
    )
